src/data/set_like.lean

```diff
@@ -0,0 +1,116 @@
+/-
+Copyright (c) 2021 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Wieser
+-/
+import data.set.basic
+
+/-!
+# Typeclass for a type `A` with an injective map to `set B`
+
+This typecalss is primary for use by subobjects like `submonoid` and `submodule`.
+
+A typical subobject should be declared as:
+```
+structure my_subobject (X : Type*) :=
+(carrier : set X)
+(op_mem : ∀ {x : X}, x ∈ carrier → sorry ∈ carrier)
+
+namespace my_subobject
+
+instance : set_like (my_subobject R) M :=
+⟨sub_mul_action.carrier, λ p q h, by cases p; cases q; congr'⟩
+
+@[simp] lemma mem_carrier {p : my_subobject R} : x ∈ p.carrier ↔ x ∈ (p : set M) := iff.rfl
+
+@[ext] theorem ext {p q : my_subobject R} (h : ∀ x, x ∈ p ↔ x ∈ q) : p = q := set_like.ext h
+
+end my_subobject
+```
+
+This file will then provide a `coe_sort`, a `coe` to set, a `partial_order`, and various
+extensionality and simp lemmas.
+
+-/
+set_option old_structure_cmd true
+
+/-- A class to indicate that there is a canonical injection between `A` and `set B`. -/
+@[protect_proj]
+class set_like (A : Type*) (B : out_param $ Type*) :=
+(coe : A → set B)
+(coe_injective' : function.injective coe)
+
+namespace set_like
+
+variables {A : Type*} {B : Type*} [i : set_like A B]
+
+include i
+
+instance : has_coe_t A (set B) := ⟨set_like.coe⟩
+
+@[priority 100]
+instance : has_mem B A := ⟨λ x p, x ∈ (p : set B)⟩
+
+-- `dangerous_instance` does not know that `B` is used only as an `out_param`
+@[nolint dangerous_instance, priority 100]
+instance : has_coe_to_sort A := ⟨_, λ p, {x : B // x ∈ p}⟩
+
+variables (p q : A)
+
+@[simp, norm_cast] theorem coe_sort_coe : ↥(p : set B) = p := rfl
+
+variables {p q}
+
+protected theorem «exists» {q : p → Prop} :
+  (∃ x, q x) ↔ (∃ x ∈ p, q ⟨x, ‹_›⟩) := set_coe.exists
+
+protected theorem «forall» {q : p → Prop} :
+  (∀ x, q x) ↔ (∀ x ∈ p, q ⟨x, ‹_›⟩) := set_coe.forall
+
+theorem coe_injective : function.injective (coe : A → set B) :=
+λ x y h, set_like.coe_injective' h
+
+@[simp, norm_cast] theorem coe_set_eq : (p : set B) = q ↔ p = q := coe_injective.eq_iff
+
+theorem ext' (h : (p : set B) = q) : p = q := coe_injective h
+
+theorem ext'_iff : p = q ↔ (p : set B) = q := coe_set_eq.symm
+
+/-- Note: implementers of `set_like` must copy this lemma in order to tag it with `@[ext]`. -/
+theorem ext (h : ∀ x, x ∈ p ↔ x ∈ q) : p = q := coe_injective $ set.ext h
+
+theorem ext_iff : p = q ↔ (∀ x, x ∈ p ↔ x ∈ q) := coe_injective.eq_iff.symm.trans set.ext_iff
+
+variables (p)
+@[simp] theorem mem_coe {x : B} : x ∈ (p : set B) ↔ x ∈ p := iff.rfl
+variables {p}
```

Comment:
Why is `p` an explicit variable here? `p` should be inferrable from the context, and it looks like most of the usages (and most of the previous `mem_coe`s) leave it implicit.

Reply:
It's explicit because it's explicit on [`submodule.mem_coe`](https://leanprover-community.github.io/mathlib_docs/algebra/module/submodule.html#submodule.mem_coe), and I figured that `submodules` were used more frequently than any other subobject.

Reply:
On the other hand, the `submodule` API is slightly outdated and has more definitions with too many explicit parameters (for my taste). I think we should follow the lead of newer definitions like `subring`, which were made with the experiences of `submodule` in hindsight.

Reply:
I guess the primary motivation was probably dot notation, which is of course moot now that it's in the `set_like` namespace.

I'm worried it will be a lot of work to make this change, but I'll have a go at just doing a find/replace for `(set_like.mem_coe _)` and see if that gets them all

Reply:
Find and replace was enough!